File: app/src/main/java/com/example/theexchange/data/store/remote/api/ApiManager.java
package com.example.theexchange.data.store.remote.api;

import retrofit2.Retrofit;


public class ApiManager {

    //Default endpoint
    private static ApiManager apiManager;

    private Retrofit retrofit;

    private static String ENDPOINT = "";


    private ApiManager(String apiEndpoint, boolean encrypted, Integer timeout) {
        retrofit = ApiAdapterFactory.createAdapter(apiEndpoint, timeout);
    }

    public static synchronized ApiManager getApiInstance() {
        if (apiManager == null) {
            apiManager = new ApiManager(ENDPOINT, false, null);
        }
        return apiManager;
    }

}

File: app/src/main/java/com/example/theexchange/data/store/remote/api/ApiManager.kt
package com.example.theexchange.data.store.remote.api

import com.example.theexchange.BuildConfig
import com.example.theexchange.data.store.remote.api.service.CountryService
import com.example.theexchange.data.store.remote.api.service.ExchangeService
import retrofit2.Retrofit


class ApiManager private constructor(apiEndpoint: String, timeout: Int?) {

    private val retrofit: Retrofit = ApiAdapterFactory.createAdapter(apiEndpoint, timeout)

    private var countryService: CountryService? = null

    private var exchangeService: ExchangeService? = null

    fun getCountryService(): CountryService {
        if (countryService == null) {
            countryService = retrofit.create(CountryService::class.java)
        }
        return countryService!!
    }

    fun getExchangeService(): ExchangeService {
        if (exchangeService == null) {
            exchangeService = retrofit.create(ExchangeService::class.java)
        }
        return exchangeService!!
    }

    companion object {

        private var apiManager: ApiManager? = null

        val apiInstance: ApiManager
            @Synchronized get() {
                if (apiManager == null) {
                    apiManager = ApiManager(BuildConfig.API_ENDPOINT,  null)
                }
                return apiManager!!
            }
    }

}